dafny/lcf/datalog.g4

```diff
@@ -37,26 +37,63 @@ rule
   ;
 
 clause
-  : name=Identifier '(' term_list ')'
+  : builtin
+  | app
+  | expression
+  ;
+
+builtin
+  : name=Builtin '(' term_list ')'
+  ;
+
+app
+  : name=Identifier ( '(' term_list ')' )?
+  ;
+
+expression
+  : left=term name=Operator right=term
   ;
 
 term
-  : atom
+  : constant
   | variable
   ;
 
+constant
+  : atom
+  | natural
+  | string
+  | list
+  ;
+
 clause_list
   : clause ( ',' clause)*
   ;
 
 term_list
-  : term( ',' term)*
+  : term ( ',' term)*
   ;
 
 atom
   : val=Identifier
   ;
 
+natural
+  : numeral=Int // TODO: make natural use the Nat rule, or switch support from natural numbers to integers
```

Comment:
Is there a reason we can't do this TODO now?

It's odd to have a rule called `natural` actually be an integer. I'd suggest we either rename `natural` to `integer`, or change `Int` to `Nat`. Either is fine, just the mixed naming is odd right now.

Reply:
As I mentioned in person, this may have to stay unresolved for now. To summarize, the kernel only supports natural numbers right now, but the Nat lexer rule in the parser doesn't seem to work.

Reply:
Agreed that the naming issue is worth resolving now, if we can.

Reply:
I have resolved the name conflict. The Int lexer rule has been commented out, and the rest of the Antlr uses the Nat rule.